### ISSUE TITLE
compiler bug / cast from Nil to Crystal::GenericInstanceType+ failed

### ISSUE THREAD

**Issue by sam0x17** [crystal-lang/crystal#8072]
In my attempts to monkey-patch `awscr-s3` to support 0.30.0 I came across a compiler bug:

```crystal
require "awscr-s3"
class Awscr::Signer::Header
  def <=>(other : T)
    0
  end
end
```
Error:
```
Error target [redacted] failed to compile:
cast from Nil to Crystal::GenericInstanceType+ failed, at /private/tmp/crystal-20190805-7397-smhku/crystal-0.30.0/src/compiler/crystal/semantic/abstract_def_checker.cr:215:5:215 (TypeCastError)
  from Crystal::AbstractDefChecker#find_base_generic_instantiation<Crystal::Type+, (Crystal::GenericClassType+ | Crystal::GenericModuleType)>:Crystal::GenericInstanceType+
  from Crystal::AbstractDefChecker#implements?<Crystal::Type+, Crystal::Def+, Crystal::NamedType+>:Bool
  from Crystal::AbstractDefChecker#check_implemented_in_subtypes<Crystal::NamedType+, Crystal::NamedType+, Crystal::Def+>:Nil
  from Crystal::AbstractDefChecker#check_single<Crystal::NamedType+>:Nil
  from Crystal::Program#top_level_semantic<Crystal::ASTNode+>:Tuple(Crystal::ASTNode+, Crystal::TypeDeclarationProcessor)
  from Crystal::Program#semantic<Crystal::ASTNode+, Bool>:Crystal::ASTNode+
  from Crystal::Compiler#compile<Array(Crystal::Compiler::Source), String>:Crystal::Compiler::Result
  from Crystal::Command#run:(Bool | Crystal::Compiler::Result | Nil)
  from main
Error: you've found a bug in the Crystal compiler. Please open an issue, including source code that will allow us to reproduce the bug: https://github.com/crystal-lang/crystal/issues
```

**Comment by sam0x17** [comment]
note: started from https://github.com/taylorfinnell/awscr-s3/issues/55

**Comment by asterite** [comment]
Duplicate of #8031

Already fixed in master, will be available soon in 0.30.1
